### PR TITLE
Integrate fixes, improve documentation, and create simple example for collecting 3d data for mapping

### PR DESCRIPTION
**Pull request by cpaxton**
## Motivation and Context

- add mapping code to home-robot
- examples and tests for how we can integrate ROS messages to reliably create 3d spatial semantic representations
- improve dependency setup
- fix speed issues with waiting for rgb/depth cameras
- remove duplicate code in various utils

## Changelog

- add build 3d map code

## How Has This Been Tested

- robot tests

## Types of changes

- [X] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: example/build_3d_map.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+import argparse
+import sys
+import timeit
+
+import click
+import numpy as np
+import open3d
```
Comment (srama2512) [inline example/build_3d_map.py:11]: @cpaxton - import open3d as o3d can fail on some older CPUs. It's worth making this import optional (similar to cv2 imports in habitat-lab: https://github.com/facebookresearch/habitat-lab/blob/cc6c8065d1018b45f19f56a84384e4407542a2cb/habitat-lab/habitat/core/utils.py#L72-L91).

Comment (cpaxton) [reply example/build_3d_map.py:11]: not possible in this script unfortunately, i'm using o3d for voxelization

this won't be critical path though

**Comment by cpaxton** [comment]
Should now resolve https://github.com/facebookresearch/home-robot/issues/65

File: README.md [hunk 2]
```diff
@@ -76,6 +82,15 @@ robot.switch_to_manipulation_mode()
...
 
+### Development
+
+To develop in `home-robot`, install the git pre-commit hooks:
+```
+python -m pip install pre-commit
+cd $HOME_ROBOT_ROOT
+pre-commit install
+```
+
```
Comment (exhaustin) [inline README.md:93]: This overlaps with the "Code Contribution" section in the same README, we should combine them.
Also this goes to show that maybe our top-level README is being clogged with too much information, maybe we should move some of them (like the contact-graspnet note) to the README in src/home_robot?

File: src/home_robot/home_robot/utils/point_cloud.py [hunk 3]
```diff
@@ -15,30 +15,38 @@
...
+            print("WARNING: rgb values too high! Normalizing...")
+            rgb = rgb / np.max(rgb)
 
-    pcd = get_pcd(xyz, rgb)
+    pcd = numpy_to_pcd(xyz, rgb)
     show_pcd(pcd, orig, R, save, grasps)
 
 
-def show_pcd(pcd, orig=None, R=None, save=None, grasps=[]):
+def show_pcd(pcd: o3d.geometry.PointCloud, orig=None, R=None, save=None, grasps=[]):
```
Comment (exhaustin) [inline src/home_robot/home_robot/utils/point_cloud.py:49]: Add typing for other inputs. Type hints are especially helpful in utils that are meant to be used multiple times by multiple users across the codebase.

File: src/home_robot_hw/home_robot_hw/ros/camera.py [hunk 4]
```diff
@@ -49,8 +48,11 @@ def valid_pc(self, xyz, rgb, depth):
         rgb = rgb.reshape(-1, 3)[mask]
         return xyz, rgb
 
-    def wait_for_image(self):
-        rate = rospy.Rate(10)
+    def wait_for_image(self) -> None:
+        """ Wait for image. Needs to be sort of slow, in order to make sure we give it time
+        to update the image in the backend."""
+        rospy.sleep(0.5)
+        rate = rospy.Rate(2)
```
Comment (exhaustin) [inline src/home_robot_hw/home_robot_hw/ros/camera.py:55]: Use class parameters (like `StretchEnv.goal_time_tolerance`), or constants defined at top of file.

**Review by exhaustin — APPROVED** [review]
Generally looks good, some minor comments.
Thanks!

File: src/home_robot_hw/home_robot_hw/ros/camera.py (hunk 4 above)
Comment (zephirefaith) [reply src/home_robot_hw/home_robot_hw/ros/camera.py:55]: +1 to informative constant name in file

**Review by zephirefaith — APPROVED** [review]
Looks good. Do remove the line with the broken import, even if you don't address the other comments, otherwise it is fine to merge in.

File: README.md [hunk 5]
```diff
@@ -9,6 +9,18 @@
 
 Mostly Hello Stretch infrastructure
 
+## Core Concepts
+
+This package assumes you have a low-cost mobile robot with limited compute -- initially a [Hello Robot Stretch](hello-robot.com/) - and a "server" workstation with more GPU compute. Both are assumed to be running on the same network.
+
+In general this is the recommended workflow:
```
[the remark this answers is not in the document]
Comment (cpaxton) [reply README.md:16]: Thanks!

File: README.md (hunk 2 above)
Comment (cpaxton) [reply README.md:93]: I agree, let's file an issue to redo our documentation?
i think the top-level is pretty out of date, but agree we need a minimal setup

File: src/home_robot_hw/home_robot_hw/env/stretch_abstract_env.py [hunk 6]
```diff
@@ -262,13 +262,13 @@ def _create_pubs_subs(self):
...
             timeout=rospy.Duration(30.0)
         )
-        print("... connected.")
         if not server_reached:
             print("ERROR: Failed to connect to arm action server.")
             rospy.signal_shutdown(
                 "Unable to connect to arm action server. Timeout exceeded."
             )
             sys.exit()
+        print("... connected.")
```
[the remark this answers is not in the document]
Comment (cpaxton) [reply src/home_robot_hw/home_robot_hw/env/stretch_abstract_env.py:272]: added

File: src/home_robot/home_robot/utils/point_cloud.py (hunk 3 above)
Comment (cpaxton) [reply src/home_robot/home_robot/utils/point_cloud.py:49]: added

File: projects/collect_data_real_robot.py [hunk 7]
```diff
@@ -14,6 +14,7 @@
 from home_robot.agent.perception.constants import coco_categories
 from home_robot.hw.ros.stretch_ros import HelloStretchROSInterface
```
[the remark this answers is not in the document]
Comment (cpaxton) [reply projects/collect_data_real_robot.py]: done

File: README.md (hunk 2 above)
Comment (cpaxton) [reply README.md:93]: https://github.com/facebookresearch/home-robot/issues/74

**Pull request merged**